packages/ra-core/src/sideEffect/useRedirect.ts

```diff
@@ -66,15 +73,17 @@ const useRedirect = () => {
                 // history doesn't handle that case, so we handle it by hand
                 window.location.href = redirectTo;
             } else {
-                history.push({
-                    ...parsePath(
+                navigate(
+                    parsePath(
                         resolveRedirectTo(redirectTo, basePath, id, data)
                     ),
-                    state: { _scrollToTop: true, ...state },
-                });
+                    {
+                        state: { _scrollToTop: true, ...state },
+                    }
+                );
             }
         },
-        [dispatch, history]
+        [dispatch, location, navigate]
```

Comment:
not good: the return function will be different each time we navigate, forcing to rerender every component using it. Think about buttons used outside of a tabbed form that will rerender when the user switches tabs.